GraphNG: add spanNulls config option

ryantxu opened the pull request:
FieldConfig has a nullValuesMode that we do not (yet) expose in the UI -- this addes it as a standard option in the UI.

| Flot  | React |
|---|----|
| ![nulls](https://user-images.githubusercontent.com/705951/100789596-85bbff00-33cb-11eb-95d6-21d590c58a00.gif) |  ![image](https://user-images.githubusercontent.com/705951/100808779-9e3b1200-33e9-11eb-946b-d96f0490b61e.png) |

ryantxu committed:
set null values mode in the UI

ryantxu committed:
null values to null value

leeoniya commented on packages/grafana-ui/src/utils/standardEditors.tsx:
```diff
@@ -190,6 +193,28 @@ export const getStandardFieldConfigs = () => {
+    id: 'nullValueMode',
+    path: 'nullValueMode',
+    name: 'Null values',
+    description: 'Behavior when data includes null values',
```
'Behavior for null values in the data'?

leeoniya commented on packages/grafana-data/src/types/fieldOverrides.ts:
```diff
@@ -122,4 +122,5 @@ export enum FieldConfigProperty {
   Links = 'links',
   Color = 'color',
+  NullValuesMode = 'nullValueMode',
 }
```
one is plural and the other isn't?

ryantxu replied:
now matching the actual property value

leeoniya commented on packages/grafana-ui/src/utils/standardEditors.tsx:
```diff
@@ -190,6 +193,28 @@ export const getStandardFieldConfigs = () => {
+      options: [
+        { label: 'Gaps', value: NullValueMode.Null },
+        { label: 'Connected', value: NullValueMode.Ignore, description: 'Ignore the values from any calculation' },
+        { label: 'Zero', value: NullValueMode.AsZero },
```
"Zeros" (plural to match "Gaps")

leeoniya commented on packages/grafana-ui/src/utils/standardEditors.tsx:
```diff
@@ -190,6 +193,28 @@ export const getStandardFieldConfigs = () => {
+    settings: {
+      options: [
+        { label: 'Gaps', value: NullValueMode.Null },
+        { label: 'Connected', value: NullValueMode.Ignore, description: 'Ignore the values from any calculation' },
```
"Connected" always seemed odd to me, the nulls themselves are not connected, they're "Spanned" or Ignored

torkelo replied:
Yea agree, ignored / skipped seems better

ryantxu committed:
Merge remote-tracking branch 'origin/master' into null-values-mode

ryantxu committed:
add null checks

leeoniya commented:
IMO, having `Gaps` as part of a 3-state toggle at the shared config level feels too graph/sparkline specific. on the other hand it's one place to choose the behavior for `null` values.

i would find it more intuitive perhaps if the Ignored/asZeros was general and an extra boolean toggle was added to the graph level for spanning nulls (false by default). if the shared config is set to "as zeros" then the span setting would have no visible effect (because all the nulls would have been "converted" to zeros).

dprokop committed:
Disable null value mode option in gauge bar gauge, table and stat panel

dprokop commented:
>i would find it more intuitive perhaps if the Ignored/asZeros was general and an extra boolean toggle was added to the graph level for spanning nulls (false by default). if the shared config is set to "as zeros" then the span setting would have no visible effect (because all the nulls would have been "converted" to zeros).

+1 for this. I think the null value mode describes the data behaviour rather than the aesthetics of the graph. I would separate these two things and treat them as standard field option(null value mode) and custom, graph specific toggle(gaps vs spans rendering)

ryantxu committed:
add spanNulls toggle

ryantxu committed:
add spanNulls toggle

ryantxu committed:
use radio for toggle

torkelo commented:
But then the question is should the null value mode (ignore / as zero) be a transform or standard field option? The ignore can be solved with the new filter transform

ryantxu commented:
>  (ignore / as zero) be a transform or standard field option?

It should be a transform, and we should figure out how to remove/deprecate it from reducers.  BUT i think that is out of scope for this PR ;)

ryantxu commented on packages/grafana-ui/src/utils/standardEditors.tsx:
```diff
@@ -190,6 +193,30 @@ export const getStandardFieldConfigs = () => {
+        { label: 'Null', value: NullValueMode.Null },
+        // The "connectd" of "span gaps" behavior is moved to graph settings.
+        // This option should either be a boolean or removed an just "AzZero" added as a transformer
+        // { label: 'Ignore', value: NullValueMode.Ignore, description: 'Ignore the values from any calculation' },
```
I'm hiding the "ignore" option since that has a less clear migration path.   For "AzZero" we can use a transformation

torkelo commented:
But if we hide as zero then I almost lobby skipping the null value standard field mode and only have a graph gaps mode? 

The ignore null option could in some cases be useful as an option for the reducers when calc mean etc, hmm. But is it really a standard option?? Tricky

ryantxu committed:
more logging

ryantxu committed:
Merge remote-tracking branch 'origin/master' into null-values-mode

ryantxu commented:
I just updated the title of this PR and removed the standard options config.  In a new PR I will look at how to deprecated and remove the standard `nullValueMode` -- I think that should become a transformer that would replace nulls with either 0 or the previous value

ryantxu committed:
more logging

ryantxu committed:
Merge remote-tracking branch 'origin/master' into null-values-mode

ryantxu committed:
revert gauge changes
